src/ui/public/visualize/visualization.js

```diff
@@ -0,0 +1,153 @@
+      const getVisContainer = jQueryGetter('.vis-container');
+      const getSpyContainer = jQueryGetter('.visualize-spy-container');
+
+      // Show no results message when isZeroHits is true and it requires search
+      $scope.showNoResultsMessage = function () {
+        const requiresSearch = _.get($scope, 'vis.type.requiresSearch');
+        const isZeroHits = _.get($scope,'visData.hits.total') === 0;
+        const shouldShowMessage = !_.get($scope, 'vis.params.handleNoResults');
+
+        return Boolean(requiresSearch && isZeroHits && shouldShowMessage);
+      };
+
+      const legendPositionToVisContainerClassMap = {
+        top: 'vis-container--legend-top',
+        bottom: 'vis-container--legend-bottom',
+        left: 'vis-container--legend-left',
+        right: 'vis-container--legend-right',
+      };
+
+      $scope.getVisContainerClasses = function () {
+        return legendPositionToVisContainerClassMap[$scope.vis.params.legendPosition];
+      };
+
+      $scope.spy = {};
+      $scope.spy.mode = ($scope.uiState) ? $scope.uiState.get('spy.mode', {}) : {};
+
+      const applyClassNames = function () {
+        const $visEl = getVisContainer();
+        const $spyEl = getSpyContainer();
+        if (!$spyEl) return;
+
+        const fullSpy = ($scope.spy.mode && ($scope.spy.mode.fill || $scope.fullScreenSpy));
+
+        $visEl.toggleClass('spy-only', Boolean(fullSpy));
+        $spyEl.toggleClass('only', Boolean(fullSpy));
+
+        $timeout(function () {
+          if (shouldHaveFullSpy()) {
+            $visEl.addClass('spy-only');
+            $spyEl.addClass('only');
+          }
+        }, 0);
+      };
+
+      const loadingDelay = config.get('visualization:loadingDelay');
+      $scope.loadingStyle = {
+        '-webkit-transition-delay': loadingDelay,
+        'transition-delay': loadingDelay
+      };
+
+      function shouldHaveFullSpy() {
+        const $visEl = getVisEl();
+        if (!$visEl) return;
+
+        return ($visEl.height() < minVisChartHeight)
+          && _.get($scope.spy, 'mode.fill')
+          && _.get($scope.spy, 'mode.name');
+      }
+
+      // spy watchers
+      $scope.$watch('fullScreenSpy', applyClassNames);
+
+      $scope.$watchCollection('spy.mode', function () {
+        $scope.fullScreenSpy = shouldHaveFullSpy();
+        applyClassNames();
+      });
+
+      const Visualization = $scope.vis.type.visualization;
+
+      //todo: make this not a jquery element
+      const visualization = new Visualization(getVisEl()[0], $scope.vis);
+
+
+      const renderFunction = _.debounce(() => {
+        const status = getUpdateStatus($scope);
+        visualization.render($scope.visData, status)
+          .then(() => {
+            // renderComplete
+            $scope.$emit('renderComplete');
+            if (!$scope.vis.visualizeScope) {
+              $el.trigger('renderComplete');
+            }
+          });
+        $scope.$apply();
+      }, 100);
```

Comment:
100ms seems a litte long. Pressing the apply-button introduces a noticeable delay. Could we try something shorter? I'd suggest we start with 40ms, and evaluate a little over time what the impact is.